CoreScriptsRoot/ServerCoreScripts/ClientChat/ChannelsBar.lua

```diff
@@ -0,0 +1,329 @@
+
+	if (enableRightClickToLeaveChannel) then
+		tab.NameTag.MouseButton2Click:connect(function()
+			self.LeaveConfirmationNotice.Text = string.format("Leave channel %s?", tab.ChannelName)
+			self.LeaveConfirmationFrame.LeaveTarget.Value = tab.ChannelName
+			self.LeaveConfirmationFrame:TweenPosition(UDim2.new(0, 0, 0, 0), Enum.EasingDirection.In, Enum.EasingStyle.Quad, 0.2, true)
+		end)
+	end
+
+	return tab
+end
+
+function methods:RemoveChannelTab(channelName)
+	if (not self:GetChannelTab(channelName)) then
+		error("Channel tab '" .. channelName .. "'does not exist!")
+	end
+
+	local indexName = channelName:lower() 
+	self.ChannelTabs[indexName]:Destroy()
+	self.ChannelTabs[indexName] = nil
+
+	self.NumTabs = self.NumTabs - 1
+	self:OrganizeChannelTabs()
+end
+
+function methods:GetChannelTab(channelName)
+	return self.ChannelTabs[channelName:lower()]
+end
+
+function methods:OrganizeChannelTabs()
+	local order = {}
+
+	table.insert(order, self:GetChannelTab("All"))
+	table.insert(order, self:GetChannelTab("System"))
+
+	for tabIndexName, tab in pairs(self.ChannelTabs) do
+		if (tab.ChannelName ~= "All" and tab.ChannelName ~= "System") then
+			table.insert(order, tab)
+		end
+	end
+
+	for index, tab in pairs(order) do
+		tab.GuiObject.Position = UDim2.new((index - 1) * 0.25, 0, 0, 0)
+	end
+
+	--// This does the effect of dynamic tab resizing when a player 
+	--// is in less than 4 channels. It was mathematically easier to
+	--// think about and do rather than resizing the actual tabs 
+	--// themselves.
+	local xSize = math.max(1, 4/self.NumTabs)
+	self.ScrollerFrame.Size = UDim2.new(xSize, 0, 1, 0)
+
+	self:ScrollChannelsFrame(0)
+end
+
+function methods:ScrollChannelsFrame(dir)
+	if (self.ScrollChannelsFrameLock) then return end
+	self.ScrollChannelsFrameLock = true
+
+	local newPageNum = self.CurPageNum + dir
+	if (newPageNum < 0) then
+		newPageNum = 0
+	elseif (newPageNum > 0 and newPageNum + 4 > self.NumTabs) then
+		newPageNum = self.NumTabs - 4
+	end
+
+	self.CurPageNum = newPageNum
+
+
+	local tweenTime = 0.15
+	local endPos = UDim2.new(-(self.CurPageNum * 0.25), 0, 0, 0)
+
+	self.PageLeftButton.Visible = (self.CurPageNum > 0)
+	self.PageRightButton.Visible = (self.CurPageNum + 4 < self.NumTabs)
+
+	local function UnlockFunc()
+		self.ScrollChannelsFrameLock = false
+	end
+	self.ScrollerFrame:TweenPosition(endPos, Enum.EasingDirection.InOut, Enum.EasingStyle.Quad, tweenTime, true, UnlockFunc)
+end
+
+function methods:FadeOutBackground(duration)
+	self.BackgroundTweener:Tween(duration, 1)
+end
+
+function methods:FadeInBackground(duration)
+	self.BackgroundTweener:Tween(duration, 0)
+end
+
+function methods:FadeOutText(duration)
+	self.TextTweener:Tween(duration, 1)
+end
+
+function methods:FadeInText(duration)
+	self.TextTweener:Tween(duration, 0)
+end
+
+function methods:CreateTweeners()
+	self.BackgroundTweener:CancelTween()
+	self.TextTweener:CancelTween()
+
+	self.BackgroundTweener = moduleTransparencyTweener.new()
```

Comment:
TransparencyTweener is a data structure for tweening a group of objects as a single unit. In this method CreateTweeners() we first cancel any potential pre-existing tweens, and then recreate the tweener object so we can get an empty set of group data.